pkg/kubelet/dockertools/manager.go

```diff
@@ -1119,70 +1135,81 @@ func (dm *DockerManager) KillPod(pod kubecontainer.Pod) error {
+		}
+		containerID = targetContainer.ID
 
-// TODO(vmarmol): Unexport this as it is no longer used externally.
-// KillContainer kills a container identified by containerID.
-// Internally, it invokes docker's StopContainer API with a timeout of 10s.
-// TODO: Deprecate this function in favor of KillContainerInPod.
-func (dm *DockerManager) KillContainer(containerID types.UID) error {
-	return dm.killContainer(containerID)
+	case container == nil || pod == nil:
+		// Read information about the container from labels
+		inspect, err := dm.client.InspectContainer(string(containerID))
+		if err != nil {
+			return err
+		}
+		storedPod, storedContainer, cerr := containerAndPodFromLabels(inspect)
+		if cerr != nil {
+			glog.Errorf("unable to access pod data from container: %v", err)
+		}
+		if container == nil {
+			container = storedContainer
+		}
+		if pod == nil {
+			pod = storedPod
+		}
+	}
+	return dm.killContainer(containerID, container, pod)
 }
 
-func (dm *DockerManager) killContainer(containerID types.UID) error {
+// killContainer accepts a containerID and an optional container or pod containing shutdown policies. Invoke
+// KillContainerInPod if information must be retrieved first.
+func (dm *DockerManager) killContainer(containerID types.UID, container *api.Container, pod *api.Pod) error {
 	ID := string(containerID)
-	glog.V(2).Infof("Killing container with id %q", ID)
-	inspect, err := dm.client.InspectContainer(ID)
-	if err != nil {
-		return err
+	name := ID
+	if container != nil {
+		name = fmt.Sprintf("%s %s", name, container.Name)
 	}
-	var found bool
-	var preStop string
-	if inspect != nil && inspect.Config != nil && inspect.Config.Labels != nil {
-		preStop, found = inspect.Config.Labels[kubernetesPodLabel]
+	if pod != nil {
+		name = fmt.Sprintf("%s %s/%s", name, pod.Namespace, pod.Name)
 	}
-	if found {
-		var pod api.Pod
-		err := latest.Codec.DecodeInto([]byte(preStop), &pod)
-		if err != nil {
-			glog.Errorf("Failed to decode prestop: %s, %s", preStop, ID)
-		} else {
-			name := inspect.Config.Labels[kubernetesContainerLabel]
-			var container *api.Container
-			for ix := range pod.Spec.Containers {
-				if pod.Spec.Containers[ix].Name == name {
-					container = &pod.Spec.Containers[ix]
-					break
-				}
-			}
-			if container != nil {
-				glog.V(1).Infof("Running preStop hook")
-				if err := dm.runner.Run(ID, &pod, container, container.Lifecycle.PreStop); err != nil {
-					glog.Errorf("failed to run preStop hook: %v", err)
-				}
-			} else {
-				glog.Errorf("unable to find container %v, %s", pod, name)
-			}
+
+	gracePeriod := uint(minimumGracePeriodInSeconds)
+	if pod != nil && pod.DeletionGracePeriodSeconds != nil {
+		gracePeriod = uint(*pod.DeletionGracePeriodSeconds)
+	}
+	glog.V(2).Infof("Killing container %q with %d grace period", name, gracePeriod)
+
+	if pod != nil && container != nil && container.Lifecycle != nil && container.Lifecycle.PreStop != nil {
```

Comment:
If there are multiple containers with prestop hooks within the pod, it appears we execute those hooks sequentially? Maybe we should pass in the start time, or else execute the hooks concurrently, so they all start around the same time. Otherwise, the grace periods will only be partially overlapping.